### PR TITLE
Route Based Code Splitting

### LINKED ISSUE #136: [RFC] Route Based Code Splitting
## Type of Change
- [x] New Feature Request (RFC)
- [ ] Documentation / Website
- [ ] Improvement / Suggestion
- [ ] Bug
- [ ] Other (please clarify below)

## Summary
As it stands now it looks like Greenwood only creates on _index.js_.  Would like to look into how we can possibly do route based code splitting, or I guess in the context of this project, template based code splitting?
Screen Shot 2019-09-18 at 7 59 54 AM

## Details
In particular this is a potential concern until larger projects can be tested because if ALL the content / code ends up in _index.js_ then it will absolutely crush larger projects even though currently the Greenwood website scores well.  

Ideally we should be able to configure this all through webpack.
https://webpack.js.org/plugins/split-chunks-plugin/

We should also want to handle / consider splitting out vendor code into it's own chunk too.

### PR DESCRIPTION
**Pull request by hutchgrant**
## Related Issue
Resolves #136 

## Summary of Changes
1. Modify scaffold to build lit-redux routes using lazy loaded dynamic imports
2. ~~Add babel-eslint~~
3. ~~Enable `@babel/plugin-syntax-dynamic-import` plugin in babel config~~
4. Modify webpack config to set chunk file names
5. Fix app-template test case
6. Add loading component and spinner

### CODE REVIEW

**Comment by hutchgrant** [comment]
You'll notice in the www example that since we're placing our header and footer in our page-template, we lose the SPA feel on route change with lazy loaded routes. Perhaps it might be a better idea to implement a custom app-template as part of the www as well.

Thoughts?

**Comment by hutchgrant** [comment]
Added a basic loading component w/spinner

loading

**Comment by hutchgrant** [comment]
Just a note. Given the size of our pages, I don't know if this is necessary at the moment. Though it might be a nice to have feature.

The bundle size on master is 61kb and with this PR it drops 27kb, but it has to make an extra request for an additional 5-7kb for each page that isn't the home page.  While you do reduce the bundle size by 45%, assuming you only view one single page, is it really that necessary given the small size of our pages?  And if you view more than 5-6 pages, it will actually be slower for you, than without chunks.

File: package.json [hunk 1]
```diff
@@ -24,12 +24,14 @@
     "test:tdd": "yarn clean && mocha --watch ./packages/**/test/**/**/*.spec.js --timeout 15000"
   },
   "devDependencies": {
+    "babel-eslint": "^10.0.3",
```
[the remark this answers is not in the document]
Comment (hutchgrant) [reply package.json]: It's required so that lint doesn't throw an error immediately within the editor. Yes I should move it to the package specifically.

File: packages/cli/package.json [hunk 2]
```diff
@@ -29,6 +29,7 @@
     "@babel/plugin-transform-runtime": "^7.6.2",
     "@babel/preset-env": "^7.6.0",
     "@babel/runtime": "^7.7.2",
+    "@greenwood/cli": "^0.4.2",
```
[the remark this answers is not in the document]
Comment (hutchgrant) [reply packages/cli/package.json]: Yea I removed some packages and added them, don't know how this got in there.

File: packages/cli/src/config/webpack.config.common.js [hunk 3]
```diff
@@ -87,6 +87,7 @@ module.exports = ({ config, context }) => {
     output: {
       path: path.join(context.publicDir, '.', config.publicPath),
       filename: '[name].[hash].bundle.js',
+      chunkFilename: '[name].bundle.js',
```
[the remark this answers is not in the document]
Comment (hutchgrant) [reply packages/cli/src/config/webpack.config.common.js]: of course.

File: packages/cli/src/lifecycles/scaffold.js [hunk 4]
```diff
@@ -89,30 +89,21 @@ const writePageComponentsFromTemplate = async (compilation) => {
...
-};
-
 const writeRoutes = async(compilation) => {
   return new Promise(async (resolve, reject) => {
     try {
       let data = await fs.readFile(compilation.context.appTemplatePath, 'utf8');
 
       const routes = compilation.graph.map(file => {
-        return `<lit-route path="${file.route}" component="eve-${file.label}"></lit-route>\n\t\t\t\t`;
+        return `<lit-route path="${file.route}" component="eve-${file.label}" .resolve="\${() => import(${file.relativeExpectedPath})}" loading="eve-loading"></lit-route>\n\t\t\t\t`;
```
[the remark this answers is not in the document]
Comment (hutchgrant) [reply packages/cli/src/lifecycles/scaffold.js]: That would be cool, good point. Why would a developer just want my basic spinner right?

File: package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (thescientist13) [reply package.json]: > It's required so that lint doesn't throw an error immediately within the editor. 

Not sure I understand what the editor has to do with our code though?  Does it actually cause  `eslint` to fail?

Maybe it would be better to provide a screenshot or the error message so I can see what the actual issue is, not sure this dependency is needed, at least not for the CLI since linting really shouldn't have any impact on our source code per se.

File: www/components/loading/spinner.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+  static get properties() {
+    return {
+      size: {
+        type: String
+      }
+    };
+  }
+
+  render() {
```
[the remark this answers is not in the document]
Comment (hutchgrant) [reply www/components/loading/spinner.js]: Yea that's a custom feature of lit-redux-router you're talking about. It doesn't do progressive loading.  It displays the loading component while its waiting to resolve. You could just put the loading component on a timer if you want. That's the simplest way.

**Comment by hutchgrant** [comment]
The lint errors were related to the dynamic-import-plugin which appparently is already included in our dependencies. The only way I found to get around that problem was to add a .eslintignore file and specified all node_modules/ files.

The app-template header/footer methodology is similar to any one page app. You want to keep the header/footer consistent, while the content is the only thing that lazyily loads. This is better UX IMO. Otherwise the entire page goes blank every single click and it defeats the purpose of an SPA.

File: package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (hutchgrant) [reply package.json]: yea it was highlighting the dynamic import, but I forgot I needed to make that string anyway so it doesn't matter actually, that entire dep is now not needed.

**Comment by thescientist13** [comment]
> The app-template header/footer methodology is similar to any one page app. You want to keep the header/footer consistent, while the content is the only thing that lazyily loads. This is better UX IMO. Otherwise the entire page goes blank every single click and it defeats the purpose of an SPA.

For sure, and totally understand.  I was just under the impression we were already doing something like that with this, re: `<entry>`
```
<div class='wrapper'>
  <eve-header></eve-header>

    <div class="content">
      <entry></entry>
    </div>

  <eve-footer></eve-footer>
</div>
```

But I do see now that is more of mechanism of our scaffolding, and not so much the runtime (e.g. routing), where as you point out, in a SPA, you would have something like:
```html
<header></header>
    
  <router-outlet></router-outlet>

<footer></footer>
```
With the router outlet being more a fragment / encapsulated component's HTML for _just_ page content.

An interesting distinction / challenge I see now.  Will think about some of this myself as I continue to familiarize myself with working on #115 as well, since we are now seeing similar architectural needs converging at once on this.  (as you anticipated).

**Comment by hutchgrant** [comment]
What if we set code splitting as toggleable from the greenwood config, include a default loading component, allow that to be overriden the same way we do app-template? We'd have to add logic to the context and config lifecycles for that, but it's doable.

**Comment by hutchgrant** [comment]
I've modified it to split into 4 main chunks, 1 for each of our main categories, as well as an additional vendor.js chunk. Obviously this only works for our website but I've done this as an example because you mentioned it would be cool to combine the pages based on the route.  

The tests failed because now there is an additional script element. Do you want to keep the vendor.js bundle?

File: packages/cli/src/config/webpack.config.common.js [hunk 6]
```diff
@@ -125,6 +126,60 @@ module.exports = ({ config, context }) => {
...
+        maxAsyncRequests: 6,
+        maxInitialRequests: 4,
+        automaticNameDelimiter: '~',
+        automaticNameMaxLength: 30,
+        cacheGroups: {
+          about: {
+            test(module, chunks) {
+              const regexp = /about/i;
+              const result = `${module.name || ''}`.match(regexp) || chunks.some(chunk => `${chunk.name || ''}`.match(regexp));
+
```
Comment (hutchgrant) [inline packages/cli/src/config/webpack.config.common.js]: Yes it's hardcoded into 4 cache groups, but ideally we could generate `cacheGroups` based on the graph.  This is utilizing our newly generated chunk names `page--[mainroute]--[page-name].js` which we create when the graph is filled.

Comment (hutchgrant) [reply packages/cli/src/config/webpack.config.common.js]: Or better yet, why don't we just move this to the `greenwood.config.js` and allow people to set their own chunks based on our generated graph chunk names.

File: greenwood.config.js [hunk 7]
```diff
@@ -20,6 +20,61 @@ module.exports = {
...
+    splitChunks: {
+      chunks: 'all',
+      minSize: 30000,
+      maxSize: 0,
+      minChunks: 1,
+      maxAsyncRequests: 6,
+      maxInitialRequests: 4,
+      automaticNameDelimiter: '~',
+      automaticNameMaxLength: 30,
+      cacheGroups: {
```
Comment (thescientist13) [inline greenwood.config.js]: yeah, We should be doing this in _webpack.config.common.js_ for the user by default.

And also do some sort of dynamic creation of `cacheGroups` here like we do in _webpack.config.common.js_ with `NormalModuleReplaclementPlugin`.

File: packages/cli/src/lifecycles/config.js [hunk 8]
```diff
@@ -11,6 +11,7 @@ let defaultConfig = {
   publicPath: '/',
   title: 'Greenwood App',
   meta: [],
+  optimization: {},
```
Comment (thescientist13) [inline packages/cli/src/lifecycles/config.js]: this should not be a public API (yet), we'll think of something but I want to keep this under the hood

File: greenwood.config.js (hunk 7 above)
Comment (hutchgrant) [reply greenwood.config.js]: I was thinking along the lines of similar to a plugin, if a developer wants custom split routes they can configure it however they please.

File: packages/cli/src/lifecycles/scaffold.js [hunk 9]
```diff
@@ -89,30 +89,21 @@ const writePageComponentsFromTemplate = async (compilation) => {
 
 };
 
-const writeListImportFile = async (compilation) => {
```
[the remark this answers is not in the document]
Comment (thescientist13) [reply packages/cli/src/lifecycles/scaffold.js:74]: Does this [get rid of `MDIMPORT`](https://github.com/ProjectEvergreen/greenwood/issues/322)!?

File: www/components/loading/spinner.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (thescientist13) [reply www/components/loading/spinner.js:13]: I think let's back this work out then and open it as a separate PR for us to discuss further in, just so we can evaluate on its own and then it makes this PR a bit "cleaner" for review purposes.  I think there is something interesting here, but let's take it to another thread.  👍

File: www/templates/app-template.js [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+        header.replaceChild(meta, oldmeta);
+      } else {
+        header.appendChild(meta);
+      }
+    });
+  }
+
+  render() {
+    return html`
+      <div class='wrapper'>
```
Comment (thescientist13) [inline www/templates/app-template.js:96]: Seems a shame we need a custom _app-template.js_ just for a few lines of HTML.  🤔 

I wonder if we could inherit from Greenwoood's `AppComponent` and just override just the `render` method somehow???  That would be pretty slick...

File: www/templates/app-template.js [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+        <eve-header></eve-header>
+        MYROUTES
```
Comment (thescientist13) [inline www/templates/app-template.js:98]: Is `MYROUTES` new?  Would be a shame to have add another "magic constants" when we're trying to move away from that (`MDIMPORT`, `META****`... any other thoughts on how to approach this?

File: packages/cli/src/config/webpack.config.common.js [hunk 12]
```diff
@@ -138,6 +138,8 @@ module.exports = ({ config, context }) => {
       }]
     },
 
+    optimization: config.optimization,
```
Comment (thescientist13) [inline packages/cli/src/config/webpack.config.common.js:141]: let's back out exposing this to the user for now
https://github.com/ProjectEvergreen/greenwood/pull/265/files#r355764479

File: packages/cli/src/lifecycles/graph.js [hunk 13]
```diff
@@ -76,6 +76,16 @@ const createGraphFromPages = async (pagesDir, config) => {
                 // set <title></title> element text, override with markdown title
                 title = title || '';
 
+                // create webpack chunk name based on route and page name
+                const routes = route.lastIndexOf('/') === route.length - 1 && route.lastIndexOf('/') > 0
```
Comment (thescientist13) [inline packages/cli/src/lifecycles/graph.js:80]: what is this conditional expressing?  Maybe better to put it into a variable to improve readability?

File: packages/cli/src/lifecycles/graph.js [hunk 14]
```diff
@@ -76,6 +76,16 @@ const createGraphFromPages = async (pagesDir, config) => {
...
+                  ? route.substring(1, route.length - 1).split('/')
+                  : route.substring(1, route.length).split('/');
+                let chunkName = 'page';
+
+                routes.map(subDir => {
+                  chunkName += '--' + subDir;
```
Comment (thescientist13) [inline packages/cli/src/lifecycles/graph.js:86]: if we're not actually mapping anything, a `forEach` is probably more appropriate.  And we can also use a template string literal as well
```js
let chunkName = '';

routes.forEach(subDir => {
  chunkName = `page--${subDir}`;
});
```

Also, if this generating a lit-route (presumably) how come we are getting an array of routes?

File: packages/cli/src/templates/app-template.js [hunk 15]
```diff
@@ -59,7 +58,7 @@ class AppComponent extends LitElement {
     meta.forEach(metaItem => {
       const metaType = metaItem.rel // type of meta
         ? 'rel'
-        : metaItem.name 
+        : metaItem.name
```
Comment (thescientist13) [inline packages/cli/src/templates/app-template.js:61]: i get the page was updated, but unnecessary formatting.

File: www/templates/app-template.js (hunk 11 above)
Comment (hutchgrant) [reply www/templates/app-template.js:98]: MYROUTES isn't new

Comment (thescientist13) [reply www/templates/app-template.js:98]: ah yes, i missed that one.  thanks for catching me on that 👍

Comment (thescientist13) [reply www/templates/app-template.js:98]: I will add it tracking on that issue then.  Hopefully we can find a "clean" way around those as well.

File: packages/cli/src/lifecycles/scaffold.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (hutchgrant) [reply packages/cli/src/lifecycles/scaffold.js:74]: No it gets rid of the ./list.js which is a list of components to import in the app-template

File: packages/cli/src/lifecycles/graph.js (hunk 14 above)
Comment (hutchgrant) [reply packages/cli/src/lifecycles/graph.js:86]: No weren't getting an array, we're stepping through a route and getting each directory to put it as part of the chunk name.  We're splitting the route by `/` and then placing it into one string with `chunk += 'page--' + route`

e.g. /docs/markdown  ===   page--docs--markdown

File: packages/cli/src/lifecycles/graph.js (hunk 13 above)
Comment (hutchgrant) [reply packages/cli/src/lifecycles/graph.js:80]: becausewedwouldhavevariablenamesthislong

Comment (thescientist13) [reply packages/cli/src/lifecycles/graph.js:80]: I only asked a question 🤷

File: packages/cli/src/lifecycles/graph.js (hunk 14 above)
Comment (thescientist13) [reply packages/cli/src/lifecycles/graph.js:86]: Gotcha, thanks for clarifying.

And just to follow up, is there anything meaningful to this this string format, per se?
```
page--docs--markdown
```

Or just a convention you came up with?

File: packages/cli/src/lifecycles/graph.js (hunk 13 above)
Comment (hutchgrant) [reply packages/cli/src/lifecycles/graph.js:80]: I mean I originally wrote that 6 months ago and was still able to understand it fairly quickly.

File: packages/cli/src/lifecycles/graph.js (hunk 14 above)
Comment (hutchgrant) [reply packages/cli/src/lifecycles/graph.js:86]: just a convention I came up with, I think I've seen something similar elsewhere.

File: packages/cli/src/lifecycles/graph.js (hunk 13 above)
Comment (thescientist13) [reply packages/cli/src/lifecycles/graph.js:80]: Yeah, I'm just trying to think more of how this would be read / understood by someone else coming to the project fresh who hasn't spent a year+ working on it, or how you would explain the code to someone else as you walked through it.  Length or mild verboseness shouldn't really be considered as some sort of gatekeeping metric, to be fair, but rather the value it adds to the overall expressiveness of the code.  Mapping logic to context (variables) can thus help reinforce the intent / readability of the code so much so that it almost comes out as natural language (or at least is always my goal).

When the code itself reads like a book, then you never really need comments, unless it's maybe capture the why (specific business rule / case, environment constraint, etc), as opposed to the what.

Another factor to my comment is also more that we have a lot of similar logic like this around our code [[1](https://github.com/ProjectEvergreen/greenwood/blob/master/packages/cli/src/lifecycles/graph.js#L45)], [[2](https://github.com/ProjectEvergreen/greenwood/blob/master/packages/cli/src/lifecycles/scaffold.js#L45)], like in our resolvers and serialize code that I think we have come to accept over time.  I have a card in Trello where I had wanted to make a more technical document for ourselves and contributing developers that would explain the architecture of the project, and how it is organized overall.  I think a key outcome of that exercise will be a better understood taxonomy (or in Domain Driven Design, they would call it a ubiquitous language) which would then allow us to better express these (at first glance) arbitrary bits of logic, and / or could lead to a more refined implementation of Greenwood itself as we get more awareness of what we're trying to achieve.

I'll add this convo as notes to that card.

Comment (thescientist13) [reply packages/cli/src/lifecycles/graph.js:80]: hah.  there was already a note for this exact topic in the card already 🤓

File: packages/cli/src/lifecycles/scaffold.js [hunk 16]
```diff
@@ -71,30 +71,25 @@ const writePageComponentsFromTemplate = async (compilation) => {
...
     try {
       let data = await fs.readFile(compilation.context.appTemplatePath, 'utf8');
 
       const routes = compilation.graph.map(file => {
-        return `<lit-route path="${file.route}" component="eve-${file.label}"></lit-route>\n\t\t\t\t`;
+        return `<lit-route
+          path="${file.route}"
+          component="eve-${file.label}"
+          .resolve="\${() => import(/* webpackChunkName: "${file.chunkName}" */ ${file.relativeExpectedPath})}"
+          ></lit-route>\n\t\t\t\t`;
```
Comment (thescientist13) [inline packages/cli/src/lifecycles/scaffold.js:84]: what's the `\n\t\t\t\t` for?  And foe the first `$` need to be escaped?

Comment (hutchgrant) [reply packages/cli/src/lifecycles/scaffold.js:84]: just for code formatting, not necessary to function.

Comment (thescientist13) [reply packages/cli/src/lifecycles/scaffold.js:84]: Not sure I follow.  Formatting in the rendered HTML?   Not sure that has ever really been a concern?

File: www/templates/app-template.js [hunk 17]
```diff
@@ -23,6 +23,13 @@ connectRouter(store);
 
 class AppComponent extends LitElement {
 
+  static get styles() {
```
Comment (thescientist13) [inline www/templates/app-template.js:26]: 👌

File: packages/cli/src/lifecycles/scaffold.js (hunk 16 above)
Comment (thescientist13) [reply packages/cli/src/lifecycles/scaffold.js:84]: Ah, I see.  


Screen Shot 2020-06-03 at 8 49 40 PM
It is the difference between this (with)


and this (without)
Screen Shot 2020-06-03 at 8 49 49 PM

I don't think it's a big deal, and we can probably omit it.  I think if anything, we should / could use a html minfier on the final output just save even more bytes.  I'll handle removing the formatting and tracking html minification.
